File: typescript/packages/management/src/tools/create_agent.ts
import {
  DEFAULT_MODEL_BY_BACKEND,
  type AgentBackend,
  type AgentModelSettings,
  type CreateAgentRequest,
} from '@copass/core';

import type { ToolContext, ToolHandler } from '../registrar.js';

export const createAgent: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const backend: AgentBackend =
    input.backend === 'google' ? 'google' : 'anthropic';
  const defaultModel = DEFAULT_MODEL_BY_BACKEND[backend];

  const modelSettings: AgentModelSettings = {
    backend,
    model: typeof input.model === 'string' ? input.model : defaultModel,
  };
  if (typeof input.temperature === 'number') {
    modelSettings.temperature = input.temperature;
  }
  if (typeof input.max_tokens === 'number') {
    modelSettings.max_tokens = input.max_tokens;
  }
  if (typeof input.max_turns === 'number') {
    modelSettings.max_turns = input.max_turns;
  }
  if (typeof input.timeout_s === 'number') {
    modelSettings.timeout_s = input.timeout_s;
  }

  const request: CreateAgentRequest = {
    slug: String(input.slug),
    name: String(input.name),
    system_prompt: String(input.system_prompt),
    model_settings: modelSettings,
  };
  if (typeof input.description === 'string') {
    request.description = input.description;
  }
  if (Array.isArray(input.tool_allowlist)) {
    request.tool_allowlist = input.tool_allowlist.map(String);
  }

  const agent = await ctx.client.agents.create(ctx.sandboxId, request);
  return { agent };
};

File: python/copass-management/src/copass_management/tools/create_agent.py
from __future__ import annotations

from typing import Any, Dict, TYPE_CHECKING

from copass_core import DEFAULT_MODEL_BY_BACKEND

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def create_agent(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    backend = str(input.get("backend") or "anthropic")
    default_model = DEFAULT_MODEL_BY_BACKEND.get(
        backend, DEFAULT_MODEL_BY_BACKEND["anthropic"]
    )
    model_settings: Dict[str, Any] = {
        "backend": backend,
        "model": str(input.get("model") or default_model),
    }
    if "temperature" in input:
        model_settings["temperature"] = float(input["temperature"])
    if "max_tokens" in input:
        model_settings["max_tokens"] = int(input["max_tokens"])
    if "max_turns" in input:
        model_settings["max_turns"] = int(input["max_turns"])
    if "timeout_s" in input:
        model_settings["timeout_s"] = int(input["timeout_s"])

    kwargs: Dict[str, Any] = {
        "slug": str(input["slug"]),
        "name": str(input["name"]),
        "system_prompt": str(input["system_prompt"]),
        "tool_allowlist": list(input.get("tool_allowlist") or []),
        "model_settings": model_settings,
    }
    if "description" in input and input["description"] is not None:
        kwargs["description"] = str(input["description"])

    agent = await ctx.client.agents.create(ctx.sandbox_id, **kwargs)
    return {"agent": agent}
